File: src/domain/services/deck.ts
import type { Rank } from '../value-objects/rank';
import { RANKS } from '../value-objects/rank';
import { SUITS } from '../value-objects/suit';
import { Card } from '../value-objects/card';

export type DealtHands = {
  viraCard: Card;
  playerOneHand: Card[];
  playerTwoHand: Card[];
};

function buildDeck(): Card[] {
  const deck: Card[] = [];

  for (const rank of RANKS) {
    for (const suit of SUITS) {
      deck.push(Card.from(`${rank}${suit}`));
    }
  }

  return deck;
}

function shuffle(cards: Card[]): Card[] {
  const deck = [...cards];

  for (let index = deck.length - 1; index > 0; index -= 1) {
    const swapIndex = Math.floor(Math.random() * (index + 1));

    const current = deck[index]!;
    deck[index] = deck[swapIndex]!;
    deck[swapIndex] = current;
  }

  return deck;
}

export function dealHandsFromViraRank(viraRank: Rank): DealtHands {
  const shuffledDeck = shuffle(buildDeck());

  const viraIndex = shuffledDeck.findIndex((card) => card.getRank() === viraRank);
  if (viraIndex < 0) {
    throw new Error(`Could not find a vira card for rank ${viraRank}.`);
  }

  const [viraCard] = shuffledDeck.splice(viraIndex, 1);

  if (!viraCard) {
    throw new Error('Failed to extract vira card from deck.');
  }

  const playerOneHand = shuffledDeck.splice(0, 3);
  const playerTwoHand = shuffledDeck.splice(0, 3);

  if (playerOneHand.length !== 3 || playerTwoHand.length !== 3) {
    throw new Error('Failed to deal three cards to both players.');
  }

  return {
    viraCard,
    playerOneHand,
    playerTwoHand,
  };
}

File: python-bot-service/simulation/deck.py
from dataclasses import dataclass
from random import Random

RANKS = ('4', '5', '6', '7', 'Q', 'J', 'K', 'A', '2', '3')
SUITS = ('C', 'O', 'P', 'E')
FULL_DECK = tuple(f'{rank}{suit}' for rank in RANKS for suit in SUITS)


@dataclass(frozen=True)
class Deal:
    vira_card: str
    vira_rank: str
    player_one_hand: list[str]
    player_two_hand: list[str]


def deal(rng: Random) -> Deal:
    cards = list(FULL_DECK)
    rng.shuffle(cards)

    vira_card = cards.pop(0)
    player_one_hand = cards[:3]
    player_two_hand = cards[3:6]

    return Deal(
        vira_card=vira_card,
        vira_rank=vira_card[:-1],
        player_one_hand=player_one_hand,
        player_two_hand=player_two_hand,
    )
